contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -942,12 +942,27 @@ transform_likenode(Node *node)
 		 * by a user. So, it is safe to go ahead with replacing the ci_as
 		 * collation with a corresponding cs_as one if an ILIKE node is found
 		 * during dump and restore.
+		 * For CS_AI collation, we keep the LIKE operator but convert it to
+		 * CS_AS by adding a call to remove_accents_internal*. During restore,
+		 * it will take the same code path and will try to add a new call to
+		 * remove_accents_internal* on top of current node. So, we take care
+		 * of that case here.
 		 */
 		init_and_check_collation_callbacks();
-		if ((*collation_callbacks_ptr->has_ilike_node) (node) && babelfish_dump_restore)
+		if (((*collation_callbacks_ptr->has_ilike_node) (node) || (*collation_callbacks_ptr->has_like_node) (node)) && babelfish_dump_restore)
 		{
 			int			collidx_of_cs_as;
 
+			/* 
+			 * We only need to update the collateflags for CS_AI to CS_AS
+			 * No need to check whether ilike or like is present as
+			 * ilike comes only for CI collations. 
+			 * So, we know for sure that if this is CS_AI, only LIKE node 
+			 * will be present, hence we can update.
+			 */
+			if (coll_info_of_inputcollid.collateflags == 0x000e)
+				coll_info_of_inputcollid.collateflags = 0x000c;
+
```

Comment:
Changing flags just to avoid following if condition is really not needed. Lets refactor this code a bit. 

Leave original code for ILIKE as it is and introduce new piece of code above that to handle our requirement and return early from this function.

Reply:
There are 2 things to consider here:
1. Provide the correct collation information to OpExpr node
2. Need to update the collateflags
The first point is needed to have the correct collation information. For our case, it will be CS_AS as for all the other collation we update the underlying node in that way. Till now we were taking care of this for ILIKE, which means CI collations were getting updated with correct information. 
But for LIKE, which means CS_AI (we do not modify the operator) it does not pick correct information. So we need to update that anyway.

Secondly, the collateflags needs to updated as we are setting the collation information to CS_AS and we do not want to go through the handling of CS_AI collation for LIKE. We handle different collation based on collateflags and whether the operator is LIKE or not. So in the first step we have the correct meta-data, in the second step we update the collateflags to correct collateflag, else it will be CS_AI which is not expected.